### PR TITLE
KLIP-17: SQL Union

### PR DESCRIPTION
**Pull request by big-andy-coates**
### Description 

KLIP for replacing `INSERT INTO` with `UNION ALL`.

Some of the internal details still need working out, but this outlines the concepts and is a good starting point for discussion.

Easy to view version can be found [here](https://github.com/confluentinc/ksql/blob/095dc24c5cdf2eba54b8dc6dcf725123a3f28c2d/design-proposals/klip-17-sql-union.md)

### CODE REVIEW

**Comment by PeterLindner** [comment]
@big-andy-coates I hope I don't annoy you mentioning this topic again, but how do you feel about the ordering of the query output?

If multiple merged streams generate messages with the same key and the ordering of the output is not guaranteed, processing historical data would yield a different result each time it is processed.

I guess reading from all the input streams could be synchronized similar to joins, but that requires all the streams to be co-partitioned, right?

What do you think? Is this even a problem?

File: design-proposals/klip-17-sql-union.md [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+1. It reduced our ability to optimise the topologies and state stores used to build `all_orders`: 
+e.g. if each SELECT writing to `all_orders` was join with some reference table, then each KS topology 
+would have a duplicate state-store holding the materialized reference table. 
+
+The first few drawbacks directly effect our users, making the conceptual model harder to understand 
+and operating KSQL more tiresome, and potentially hurt adoption.  
+
+The later drawbacks are stopping other improvements to KSQL. Including, though not limited to:
+
+* [KLIP-18](klip-18-distributed-metastore .md): Distributed Metastore: which requires the (1 - 0..1)
```
Comment (rodesai) [inline design-proposals/klip-17-sql-union.md]: could you add a note on why this is true?

**Comment by rodesai** [comment]
> I guess reading from all the input streams could be synchronized similar to joins, but that requires all the streams to be co-partitioned, right?
What do you think? Is this even a problem?

I think this is out of scope for this work, but this does put us in better position to give this kind of guarantee. Before, we had n topologies for n queries. Now that everything would be in 1 topology, we could actually implement synchronization across the streams.

**Comment by big-andy-coates** [comment]
@PeterLindner 

> @big-andy-coates I hope I don't annoy you mentioning this topic again, but how do you feel about the ordering of the query output?

Not at all.

> If multiple merged streams generate messages with the same key and the ordering of the output is not guaranteed, processing historical data would yield a different result each time it is processed.
>
> I guess reading from all the input streams could be synchronized similar to joins, but that requires all the streams to be co-partitioned, right?
> 
> What do you think? Is this even a problem?

I would imagine, though I'd need to confirm, that KS will merge the streams with a best-efforts on aligning event-time across the streams. cc @mjsax 

However, order is less important IMHO when interleaving messages from multiple streams, as a stream does not have the per-key update semantics of a table.  The order from each source will be maintained in the output, but the interleaving of multiple sources will not.  

Ensuring exact ordering is tricky as a stream never ends, so there is always the possibility of out-of-order data coming, meaning you'd have to buffer forever to ensure correct ordering.  Correct event time ordering is therefore a trade off between correctness and latency. This is not something we expose in KSQL.

**Comment by big-andy-coates** [comment]
@rodesai 

> Took a first pass through. I think, as you say, the main thing we'd lose by dropping insert into is the ability to add queries writing to a sink. But we can reframe that as evolving the old query. That's non-trivial, but I agree that what's proposed here gives us a simpler model and better performance. If we get feedback that adding queries is important we can look to implement that as a form of query evolution later.

Yep, totally.

File: design-proposals/klip-17-sql-union.md [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+
+Resulting in a `all_orders` stream that has a schema that matches both `SELECT`s, and contains all 
+rows from both streams.
+
+However, this approach has several drawbacks.
+
+1. Each additional stream that needs merging into `all_orders` requires a new SQL statement
+and underlying KS topology.
+1. The is no conceptual reason that the first statement should be any different from any other
+statement inserting into `all_orders`, yet the first must be a `CSAS`, while the latter `INSERT INTO`.
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply design-proposals/klip-17-sql-union.md]: Agreed. But its still an imbalance.

File: design-proposals/klip-17-sql-union.md [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+The CLI will be updated to detect `INSERT INTO` statements and inform the user they are no longer supported, 
+and what to do instead, linking to more documentation. This can be done through a general pattern that can 
+be used for other removed statement types.
+
+## Test plan
+
+Suitable `QueryTranslationTest` cases will be added to cover the use of `UNION ALL`. 
+At a minimum this should replicate the QTT tests that exist for `INSERT INTO`.
+
+If [KLIP-6: Execution Plans](klip-6-execution-plans.md) has not been merged then any existing
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply design-proposals/klip-17-sql-union.md]: If its agreed, then yes!

File: design-proposals/klip-17-sql-union.md (hunk 1 above)
Comment (big-andy-coates) [reply design-proposals/klip-17-sql-union.md]: Meh... can't you just read KLIP-18? (Once I've raised it)

**Comment by hjafarpour** [comment]
`INSERT INTO` is not equivalent of `UNION ALL` and each of them have their intended purpose. We cannot replace `INSERT INTO` with `UNION ALL` and have the same behavior. For instance, in many cases we may have an existing stream and would need to add another stream to it. You can do this with `INSERT INTO` while in `UNION ALL` you have to have all queries available at the time you issue the statement. 
Why not have both, keep `INSERT INTO` and add `UNION ALL`?

**Comment by PeterLindner** [comment]
>I would imagine, though I'd need to confirm, that KS will merge the streams with a best-efforts on aligning event-time across the streams

I tried to look that up, the [Javadoc for KStream.merge](https://docs.confluent.io/current/streams/javadocs/org/apache/kafka/streams/kstream/KStream.html#merge-org.apache.kafka.streams.kstream.KStream-) says:

>There is no ordering guarantee between records from this KStream and records from the provided KStream in the merged stream. Relative order is preserved within each input stream though (ie, records within one input stream are processed in order).

>Ensuring exact ordering is tricky

I'm just asking about deterministic ordering (ie merging the same streams multiple times should yield the same ordering each time)

>However, order is less important IMHO when interleaving messages from multiple streams, as a stream does not have the per-key update semantics of a table

That's right, I guess problems only arise when multiple streams containing the same keys are merged and then fed into a state store. Imagine the following extreme situation: 

- stream1 already has a lot of messages with timestamps of the past 7 days
- stream2 is an empty stream and I create a single message now
- all of the messages of both steams have the same key
- I merged stream1 and stream2 and created stream3 out of them 7 days ago, the order of stream3 would be as expected
- If I merge stream1 and stream2 now and create stream4 out of them, the single message of stream2 would be one of the first processed messages and all messages from stream1 being processed afterwards would be considered out of date 
- If I then create the same aggregation for stream3 and stream4, I'd expect them to yield the same results, since I issued the exact same queries, but stream4 probably has a lot of records that are even out of the grace period.

Does that make sense or am I missing something?

**Comment by big-andy-coates** [comment]
> `SERT INTO` is not equivalent of `UNION ALL` and each of them have their intended purpose. We cannot replace `INSERT INTO` with `UNION ALL` and have the same behavior. For instance, in many cases we may have an existing stream and would need to add another stream to it. You can do this with `INSERT INTO` while in `UNION ALL` you have to have all queries available at the time you issue the statement.
> Why not have both, keep `INSERT INTO` and add `UNION ALL`?

The distinction you call out between `INSERT INTO` and `UNION ALL` is already called out in the KLIP itself in the `Compatibility Implications` section.

However, we will be able to extend the `UNION ALL` solution to cover this gap once we can support `UPDATE TABLE` style operations, which we're moving towards. In the meantime switching to `UNION ALL` does not, as stated, have exact parity with `INSERT INTO`.

The main driver for switching from `INSERT INTO` to `UNION ALL` is to bring about a 1-2-1 model between sink and query.  This enables us to get rid of the command topic and replace it with a distributed metastore topic / db.  This will be covered in KLIP-18 soon.

File: design-proposals/klip-17-sql-union.md [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+
+* Removal of `INSERT INTO` statements from the syntax and code. (Breaking change).
+* Introduction of `UNION ALL` support to merge multiple streams.
+* Arbitrary `SELECT` statements in the union, as long as they result in a stream, not a table.
+* Support for unions in CSAS and transient push queries.
+
+## What is not in scope
+
+* Introduction of `UNION` operator: which removes duplicates from the result set.
+We may choose to introduce this later, but is not required in any MVP.
```
Comment (mjsax) [inline design-proposals/klip-17-sql-union.md]: I don't think that "removing duplicates" is a valid concept -- each record in a _stream_ is an immutable fact that the concept of duplicates does not apply IMHO.

File: design-proposals/klip-17-sql-union.md [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+* Introduction of `UNION ALL` support on tables. though it may not be much work to add this later. 
+Support for tables is not required to remove `INSERT INTO` and unblock the other KLIPs, 
+hence it has been de-scoped.
```
Comment (mjsax) [inline design-proposals/klip-17-sql-union.md]: ksqlDB requires that table always have a primary key -- hence, I don't think it will be possible to add `UNION ALL` for tables. (Just a side remark.)

File: design-proposals/klip-17-sql-union.md [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+
+A new section will be added to the syntax reference to cover the use of `UNION ALL`. Not much beyond 
+that would be needed in our docs as this is a standard SQL operator, and hence well documented. 
+
+The release notes will include details of how to migrate `INSERT INTO` statements to use `UNION ALL`.
+
+# Compatibility Implications
+
+Removal of `INSERT ALL` is clearly a breaking change. However, we _must_ remove it if we are to unlock
+the other planned work.
```
Comment (mjsax) [inline design-proposals/klip-17-sql-union.md]: `INSERT ALL` ? Should be `INSERT INTO`?

What is unclear in general about this KLIP is, why not just add `UNION ALL` for streams, and keep `INSERT INTO` as-is? Why remove `INSERT INTO`? What other work is planned to use it?

File: design-proposals/klip-17-sql-union.md [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+
+The release notes will include details of how to migrate `INSERT INTO` statements to use `UNION ALL`.
+
+# Compatibility Implications
+
+Removal of `INSERT ALL` is clearly a breaking change. However, we _must_ remove it if we are to unlock
+the other planned work. 
+
+It is believed the proposed `UNION ALL` operator will be able to replicate all functionality currently 
+possible with `INSERT ALL`, with the exception that you will no longer be able to run additional statements
```
Comment (mjsax) [inline design-proposals/klip-17-sql-union.md]: `INSERT ALL` -> `INSERT INTO`

**Comment by mjsax** [comment]
@big-andy-coates @PeterLindner About ordering in merge. All you said is correct. Kafka Streams does not provides ordering guarantees because it's not really possible to provide them (more below). Kafka Streams also requires that both streams are co-partitioned, otherwise, it might compute the wrong result downstream (what is a known issues: https://issues.apache.org/jira/browse/KAFKA-7293). ksqlDB could (and **should!**) check for this conditions if `UNION ALL` is going to use `merge()` (what I assume).

The only "workaround" to merge streams based on timestamps would be via topics: ie, both streams must be written into a topic and a new stream is created from both as once:
```java
-- no ordering guarantee:
stream3 = stream1.merge(stream2);

-- workaround
stream1.to("topic-1");
stream2.to("topic-2");
Set<String> topics = new HashSet();
topics.add("topic-1");
topics.add("topic-2");
stream3 = builder.stream(topics);
```

Note that this "workaround" is strictly required and even if `merge()` would allow timestamp synchronized merging, it would need to implement it that way, as some "buffer" is required to temporarily block one stream. Using `max.task.idle.ms` one can configure how long a stream should be blocked... Note that infinite blocking is not an option in general.

Hence, I think it is ok as a first step to implement `UNION ALL` without any synchronization guarantees. @PeterLindner raises a fair point with his example, however, to address issue like this, corresponding grace-periods must be set downstream (at least in Kafka Streams -- ksqlDB does not offer much control atm with this regard).

For a time synchronized `UNION ALL` a new keyword could be introduced though in a follow up KLIP, if there is demand for such a feature. Rerouting add data throw topic to time-synchronize them is an expensive operation and users should have both options IMHO, and would need to explicitly pick the one or the other operator.

**Comment by PeterLindner** [comment]
With Tiered Storage, the time difference may be very high which could be inappropriate for grace periods (think years) so I'd rather consider that as a workaround.

But I don't know if this is a real problem, just wanted to bring this to your minds. I'm perfectly fine with just supporting non-syncronized `UNION ALL` for now. But I guess it should at least be mentioned in the docs.

**Comment by big-andy-coates** [comment]
Sorry for the delay in addressing feedback folks - I've been pulled onto other tasks!

> @big-andy-coates I hope I don't annoy you mentioning this topic again, but how do you feel about the ordering of the query output?
> 
> If multiple merged streams generate messages with the same key and the ordering of the output is not guaranteed, processing historical data would yield a different result each time it is processed.
> 
> I guess reading from all the input streams could be synchronized similar to joins, but that requires all the streams to be co-partitioned, right?
> 
> What do you think? Is this even a problem?

tl;dr; applying any ordering to streams is... problematic. This KLIP won't cover any ordering.

The problem with ordering guarantees is that they are hard to meet when dealing with infinite datasets.  It would be great to say KSQL guarantees ordering by event time.... but how could you implement this?  You can only output a given event once you _know_ there is going never be another event with an earlier event time coming later in the stream.  With an infinite stream you can _never_ know this without knowledge of the system architecture. 

Also, it's unclear if ksql even should be attempting to order things when talking about streams of events. Tables should be ordered: the event time is basically the last_modified column. Out-of-order updates for tables should be ignored as the newer value _replaces_ the old out-of-order one. However, this isn't true for streams.  The data can mean anything and can be ordered in any way.  I certainly don't think we should be trying to order streams by default. However, we may support ordering primitives in the syntax at some point, e.g. `ORDER BY ROWTIME`. 

SQL union has the benefit that at least all the sources are being joined into the sink within a single processing job, (as opposed to each `INSERT INTO` running in a separate job).  In theory, this makes it easier to us to provide some _best efforts_ at ordering in the future.

File: design-proposals/klip-17-sql-union.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply design-proposals/klip-17-sql-union.md]: I've added this KLIP to that project https://github.com/confluentinc/ksql/issues/4667

File: design-proposals/klip-17-sql-union.md (hunk 1 above)
Comment (big-andy-coates) [reply design-proposals/klip-17-sql-union.md]: KLIP-18 is now available viewing. Hence I'm not going to add details here.

File: design-proposals/klip-17-sql-union.md (hunk 4 above)
Comment (big-andy-coates) [reply design-proposals/klip-17-sql-union.md]: Agreed. I'll update the doc.

File: design-proposals/klip-17-sql-union.md (hunk 6 above)
Comment (big-andy-coates) [reply design-proposals/klip-17-sql-union.md]: > INSERT ALL ? Should be INSERT INTO?

updated!

> What is unclear in general about this KLIP is, why not just add UNION ALL for streams, and keep INSERT INTO as-is? Why remove INSERT INTO? What other work is planned to use it?

I thought the document was being quite clear on this. It has a section on it:

```
The later drawbacks are stopping other improvements to KSQL. Including, though not limited to:

* [KLIP-18][18]: Distributed Metastore: which requires the (1 - 0..1)
relationship between data source and persistent query that this KLIP will restore.
See the KLIP-17 section in KLIP-18 for details of why this is needed.
* [KLIP-19][19]: Introduce Materialized Views: which looks to clarify the
semantic difference between a view and a table/stream and their associated semantics. 
This KLIP is related because it removes `INSERT INTO` which inserts into an existing stream.
* [KLIP-20][20]: Remove `TERMINATE` statement: which requires the (1 - 0..1)
relationship between data source and persistent query that this KLIP will restore.
* [KLIP-21][21]: Correct 'INSERT VALUES' semantics: which depends on [KLIP-19][19].
```

**Comment by big-andy-coates** [comment]
I've added details of ordering guarantees to the KLIP.

**Comment by aabrams** [comment]
Hi, is there a plan to assign this to a version? Is this direction still the relevant one?

I would very much like to see this implemented, as together with "CREATE OR REPLACE" in 0.12 it will greatly simplify our ability to modify queries in live clusters in programmatic fashion.

**Comment by big-andy-coates** [comment]
It's not on any roadmap yet. I've been too busy elsewhere to progress this. When I get some time I'll kick start the process.  It's still relevant , etc, just needs resourcing.
